In [96]:
#importing libraries
import numpy as np
import pandas as pd

In [97]:
#reading csv
df_vaccines=pd.read_csv('vaccinations.csv')

In [98]:
df_vaccines.head(100)

location iso_code        date  total_vaccinations  people_vaccinated  \
0   Afghanistan      AFG  2021-02-22                 0.0                0.0   
1   Afghanistan      AFG  2021-02-23                 NaN                NaN   
2   Afghanistan      AFG  2021-02-24                 NaN                NaN   
3   Afghanistan      AFG  2021-02-25                 NaN                NaN   
4   Afghanistan      AFG  2021-02-26                 NaN                NaN   
..          ...      ...         ...                 ...                ...   
95  Afghanistan      AFG  2021-05-28                 NaN                NaN   
96  Afghanistan      AFG  2021-05-29                 NaN                NaN   
97  Afghanistan      AFG  2021-05-30            600152.0           480226.0   
98  Afghanistan      AFG  2021-05-31                 NaN                NaN   
99  Afghanistan      AFG  2021-06-01                 NaN                NaN   

    people_fully_vaccinated  total_boosters  daily_vaccinations_raw  \
0                       NaN             NaN                     NaN   
1                       NaN             NaN                     NaN   
2                       NaN             NaN                     NaN   
3                       NaN             NaN                     NaN   
4                       NaN             NaN                     NaN   
..                      ...             ...                     ...   
95                      NaN             NaN                     NaN   
96                      NaN             NaN                     NaN   
97                 119926.0             NaN                     NaN   
98                      NaN             NaN                     NaN   
99                      NaN             NaN                     NaN   

    daily_vaccinations  total_vaccinations_per_hundred  \
0                  NaN                             0.0   
1               1367.0                             NaN   
2               1367.0                             NaN   
3               1367.0                             NaN   
4               1367.0                             NaN   
..                 ...                             ...   
95              5907.0                             NaN   
96              5326.0                             NaN   
97              4746.0                             1.5   
98              5084.0                             NaN   
99              5102.0                             NaN   

    people_vaccinated_per_hundred  people_fully_vaccinated_per_hundred  \
0                             0.0                                  NaN   
1                             NaN                                  NaN   
2                             NaN                                  NaN   
3                             NaN                                  NaN   
4                             NaN                                  NaN   
..                            ...                                  ...   
95                            NaN                                  NaN   
96                            NaN                                  NaN   
97                            1.2                                  0.3   
98                            NaN                                  NaN   
99                            NaN                                  NaN   

    total_boosters_per_hundred  daily_vaccinations_per_million  \
0                          NaN                             NaN   
1                          NaN                            34.0   
2                          NaN                            34.0   
3                          NaN                            34.0   
4                          NaN                            34.0   
..                         ...                             ...   
95                         NaN                           147.0   
96                         NaN                           133.0   
97                       

# 1. Drop Columns

Dropping columns which are not required is one of the important data transformation as it reduces the data size for processing. I have removed daily_vaccinations_raw, daily_vaccinations, total_vaccinations_per_hundred, people_vaccinated_per_hundred, people_fully_vaccinated_per_hundred, total_boosters_per_hundred, daily_vaccinations_per_million, daily_people_vaccinated,daily_people_vaccinated_per_hundred columnns as they are not required and doesn't provide any value for this project. 

In [99]:
df_vaccines.drop(df_vaccines.iloc[:, 7:], inplace=True, axis=1)

In [100]:
df_vaccines.head()

location iso_code        date  total_vaccinations  people_vaccinated  \
0  Afghanistan      AFG  2021-02-22                 0.0                0.0   
1  Afghanistan      AFG  2021-02-23                 NaN                NaN   
2  Afghanistan      AFG  2021-02-24                 NaN                NaN   
3  Afghanistan      AFG  2021-02-25                 NaN                NaN   
4  Afghanistan      AFG  2021-02-26                 NaN                NaN   

   people_fully_vaccinated  total_boosters  
0                      NaN             NaN  
1                      NaN             NaN  
2                      NaN             NaN  
3                      NaN             NaN  
4                      NaN             NaN

# 2. Rename Columns

I have renmaed couple of columns location and iso_code to country_name and country_code as the earlier names were not clearly providing info as what they represent. Location can be city, state, country or latitude and longitude combination.

In [101]:
columns_list=['country_name','country_code','date','total_vaccinations','people_vaccinated','people_fully_vaccinated','total_boosters_administered']

In [102]:
# replacing the column names
df_vaccines.columns=columns_list
df_vaccines.head()

country_name country_code        date  total_vaccinations  \
0  Afghanistan          AFG  2021-02-22                 0.0   
1  Afghanistan          AFG  2021-02-23                 NaN   
2  Afghanistan          AFG  2021-02-24                 NaN   
3  Afghanistan          AFG  2021-02-25                 NaN   
4  Afghanistan          AFG  2021-02-26                 NaN   

   people_vaccinated  people_fully_vaccinated  total_boosters_administered  
0                0.0                      NaN                          NaN  
1                NaN                      NaN                          NaN  
2                NaN                      NaN                          NaN  
3                NaN                      NaN                          NaN  
4                NaN                      NaN                          NaN

# 3. Missing Values

Many important columns has null values or missing values as null data doesnt yeild any value, I am droping records with null values

In [103]:
df_vaccines.isnull().sum()

country_name                       0
country_code                       0
date                               0
total_vaccinations             65337
people_vaccinated              68129
people_fully_vaccinated        70814
total_boosters_administered    95015
dtype: int64

In [104]:
df_vaccines.shape

(129024, 7)

In [105]:
df_vaccines = df_vaccines.dropna()

In [106]:
df_vaccines.shape

(32058, 7)

In [107]:
df_vaccines.head(10)

country_name country_code        date  total_vaccinations  \
867       Africa     OWID_AFR  2021-10-28         189809912.0   
868       Africa     OWID_AFR  2021-10-29         193296330.0   
869       Africa     OWID_AFR  2021-10-30         194361624.0   
870       Africa     OWID_AFR  2021-10-31         196480876.0   
871       Africa     OWID_AFR  2021-11-01         197837482.0   
872       Africa     OWID_AFR  2021-11-02         199146273.0   
873       Africa     OWID_AFR  2021-11-03         199647730.0   
874       Africa     OWID_AFR  2021-11-04         202346669.0   
875       Africa     OWID_AFR  2021-11-05         202620703.0   
876       Africa     OWID_AFR  2021-11-06         204717964.0   

     people_vaccinated  people_fully_vaccinated  total_boosters_administered  
867        118429867.0               76360930.0                      28870.0  
868        120094541.0               79415052.0                      32750.0  
869        120630098.0               80078265.0                      71510.0  
870        122001362.0               80759982.0                      71510.0  
871        122691032.0               81432762.0                     124228.0  
872        123401709.0               82036961.0                     142786.0  
873        123704323.0               82251189.0                     153756.0  
874        125649818.0               83140534.0                     163744.0  
875        125793512.0               83267105.0                     173532.0  
876        126953294.0               84214931.0                     179218.0

# 4. Convert the data types

I have convertes data types of total_vaccinations, people_vaccinated, people_fully_vaccinated and total_boosters_administered to int as float doesnt have any sense to these columns. For example, There can not be 1.5 total vaccinations in a country i.e., 1.5 people getting vaccinated is wrong. 

In [108]:
df_vaccines["total_vaccinations"] = df_vaccines["total_vaccinations"].astype(int)
df_vaccines["people_vaccinated"] = df_vaccines["people_vaccinated"].astype(int)
df_vaccines["people_fully_vaccinated"] = df_vaccines["people_fully_vaccinated"].astype(int)
df_vaccines["total_boosters_administered"] = df_vaccines["total_boosters_administered"].astype(int)

In [109]:
df_vaccines.head()

country_name country_code        date  total_vaccinations  \
867       Africa     OWID_AFR  2021-10-28           189809912   
868       Africa     OWID_AFR  2021-10-29           193296330   
869       Africa     OWID_AFR  2021-10-30           194361624   
870       Africa     OWID_AFR  2021-10-31           196480876   
871       Africa     OWID_AFR  2021-11-01           197837482   

     people_vaccinated  people_fully_vaccinated  total_boosters_administered  
867          118429867                 76360930                        28870  
868          120094541                 79415052                        32750  
869          120630098                 80078265                        71510  
870          122001362                 80759982                        71510  
871          122691032                 81432762                       124228

# Unique Country name count

In [117]:
df_vaccines['country_name'].nunique()

205

# 5. Sort dataframe

Sorting the data frame to organize the data in descending order to keep the latest data at the top and in the next step we can drop other duplicate values peer country

In [110]:
df_vaccines.sort_values(by=['country_name','country_code','date','total_vaccinations'],ascending=False)

country_name country_code        date  total_vaccinations  \
129023     Zimbabwe          ZWE  2022-09-08            12161598   
129011     Zimbabwe          ZWE  2022-08-27            12145591   
129009     Zimbabwe          ZWE  2022-08-25            12133385   
129008     Zimbabwe          ZWE  2022-08-24            12121226   
129005     Zimbabwe          ZWE  2022-08-21            12105303   
...             ...          ...         ...                 ...   
871          Africa     OWID_AFR  2021-11-01           197837482   
870          Africa     OWID_AFR  2021-10-31           196480876   
869          Africa     OWID_AFR  2021-10-30           194361624   
868          Africa     OWID_AFR  2021-10-29           193296330   
867          Africa     OWID_AFR  2021-10-28           189809912   

        people_vaccinated  people_fully_vaccinated  \
129023            6416614                  4730503   
129011            6411006                  4725099   
129009            6407937                  4722200   
129008            6404253                  4719303   
129005            6400632                  4715815   
...                   ...                      ...   
871             122691032                 81432762   
870             122001362                 80759982   
869             120630098                 80078265   
868             120094541                 79415052   
867             118429867                 76360930   

        total_boosters_administered  
129023                      1014481  
129011                      1009486  
129009                      1003248  
129008                       997670  
129005                       988856  
...                             ...  
871                          124228  
870                           71510  
869                           71510  
868                           32750  
867                           28870  

[32058 rows x 7 columns]

# 6. Drop duplicates

Dropping duplicate values per country and retaining just one record per country. Count of unique country name count of 205 is matching with count of records after droppinf duplicate records. So, it shows we have achieved the target.

In [111]:
#removing duplicates
df_vaccines.drop_duplicates(subset='country_name',keep='first',inplace=True)

In [112]:
df_vaccines

country_name country_code        date  total_vaccinations  \
867          Africa     OWID_AFR  2021-10-28           189809912   
1496        Albania          ALB  2021-11-03             1926604   
2098        Algeria          DZA  2021-11-21            12008974   
2707        Andorra          AND  2021-12-12              115709   
3363         Angola          AGO  2022-03-25            17535411   
...             ...          ...         ...                 ...   
125567      Vietnam          VNM  2022-02-10           184868879   
126062        Wales     OWID_WLS  2021-10-21             4983245   
126811        World     OWID_WRL  2020-12-23             4409459   
127846        Yemen          YEM  2022-06-07              859241   
128768     Zimbabwe          ZWE  2021-12-27             7222381   

        people_vaccinated  people_fully_vaccinated  \
867             118429867                 76360930   
1496              1025180                   898625   
2098              6697263                  5298544   
2707                56628                    50551   
3363             11235059                  5993792   
...                   ...                      ...   
125567           79166489                 76057017   
126062            2421700                  2238954   
126811            2845300                    20684   
127846             692968                   441404   
128768            4105296                  3113142   

        total_boosters_administered  
867                           28870  
1496                           2799  
2098                          13167  
2707                           8530  
3363                         306560  
...                             ...  
125567                     29629103  
126062                       322591  
126811                            1  
127846                           31  
128768                         3943  

[205 rows x 7 columns]

# 7. Fix casing for columns with string data type

Converted country_name and country_code to upper case as when we combine this data with other data sources based on these columns, consistent casing will help us in combining data easily.

In [125]:
df_vaccines['country_name'] = df_vaccines['country_name'].str.upper()
df_vaccines['country_code'] = df_vaccines['country_code'].str.upper()

In [126]:
df_vaccines

country_name country_code        date  total_vaccinations  \
867          AFRICA     OWID_AFR  2021-10-28           189809912   
1496        ALBANIA          ALB  2021-11-03             1926604   
2098        ALGERIA          DZA  2021-11-21            12008974   
2707        ANDORRA          AND  2021-12-12              115709   
3363         ANGOLA          AGO  2022-03-25            17535411   
...             ...          ...         ...                 ...   
125567      VIETNAM          VNM  2022-02-10           184868879   
126062        WALES     OWID_WLS  2021-10-21             4983245   
126811        WORLD     OWID_WRL  2020-12-23             4409459   
127846        YEMEN          YEM  2022-06-07              859241   
128768     ZIMBABWE          ZWE  2021-12-27             7222381   

        people_vaccinated  people_fully_vaccinated  \
867             118429867                 76360930   
1496              1025180                   898625   
2098              6697263                  5298544   
2707                56628                    50551   
3363             11235059                  5993792   
...                   ...                      ...   
125567           79166489                 76057017   
126062            2421700                  2238954   
126811            2845300                    20684   
127846             692968                   441404   
128768            4105296                  3113142   

        total_boosters_administered  
867                           28870  
1496                           2799  
2098                          13167  
2707                           8530  
3363                         306560  
...                             ...  
125567                     29629103  
126062                       322591  
126811                            1  
127846                           31  
128768                         3943  

[205 rows x 7 columns]

# Ethical Implications:

I have taken the data from https://ourworldindata.org/covid-vaccinations. We will have below ethical implications:
1. As stated by website in their legal disclaimer, the can have errors, might not be accurate.
2. As we are accessing the data from the website which is researched by a small group of reseachers, we are using third-party data.
3. As I have gone through this website multiple times to confirm the data, this would have performance issues on the website.
4. As the data might have collected from multiple sources (which are unknown), data accuracy is a concern.
5. As we are unsure of the source of the data, we have issues with transparency.
6. As this is vaccination data and is available on internet at multiple places, I don't think there is a legal concern.

In short, We might have issues with legal concerns, however, we can confirm that we will have data accuracy, lack of transparency, performance issues of website.

# Steps Completed:

I have completed below steps:
    1. Drop Columns
    2. Rename Columns
    3. Drop missing values
    4. Convert data types
    5. Sorting
    6. Drop duplicates
    7. Fix casing for columns with string data type